onnx/shape_inference/implementation.h

```diff
@@ -241,6 +241,131 @@ struct InferenceContextImpl : public InferenceContext {
+
+    for (auto& attr : *n.mutable_attribute()) {
+      attributesByName_[attr.name()] = &attr;
+      if (attr.has_g()) {
+        // need a mutable GraphProto to run inferencing on this attribute
+        graphProtoAttributesByName_[attr.name()] = attr.mutable_g();
+      }
+    }
+
+    size_t input_idx = 0;
+    for (const auto& input : n.input()) {
+      inputIndexToNameMap_.insert({input_idx++, input});
+
+      auto valueTypesIter = valueTypesByName.find(input);
+      if (valueTypesIter != valueTypesByName.end()) {
+        allInputTypes_.push_back(valueTypesIter->second);
+      } else {
+        allInputTypes_.push_back(nullptr);
+      }
+
+      const auto inputDataIter = inputDataByName.find(input);
+      if (inputDataIter != inputDataByName.cend()) {
+        allInputData_.push_back(inputDataIter->second);
+      } else {
+        allInputData_.push_back(nullptr);
+      }
+    }
+
+    size_t output_idx = 0;
+    for (const auto& output : n.output()) {
+      outputIndexToNameMap_.insert({output_idx++, output});
+    }
+
+    allOutputTypes_.resize(n.output_size());
+  }
+
+  std::vector<const TensorProto*> allInputData_;
+  std::unordered_map<size_t, std::string> inputIndexToNameMap_;
+  std::unordered_map<size_t, std::string> outputIndexToNameMap_;
+  std::vector<const TypeProto*> allInputTypes_;
+  std::vector<TypeProto> allOutputTypes_;
+  std::unordered_map<std::string, TensorShapeProto>& generatedShapeData_;
+  const GraphInferenceContext* graphInferenceContext_;
+  std::unordered_map<std::string, const AttributeProto*> attributesByName_;
+  std::unordered_map<std::string, GraphProto*> graphProtoAttributesByName_;
+
+  const AttributeProto* getAttribute(const std::string& name) const override {
+    auto iter = attributesByName_.find(name);
+    if (iter == attributesByName_.end()) {
+      return nullptr;
+    } else {
+      return iter->second;
+    }
+  }
+
+  size_t getNumInputs() const override {
+    return allInputTypes_.size();
+  }
+
+  const TypeProto* getInputType(size_t index) const override {
+    if (index >= allInputTypes_.size()) {
+      ONNX_THROW("Input " + ONNX_NAMESPACE::to_string(index) + " is out of bounds.");
+    }
+    return allInputTypes_[index];
+  }
+
+  size_t getNumOutputs() const override {
+    return allOutputTypes_.size();
+  }
+
+  TypeProto* getOutputType(size_t index) override {
+    if (index >= allOutputTypes_.size()) {
+      ONNX_THROW("Output " + ONNX_NAMESPACE::to_string(index) + " is out of bounds.");
+    }
+    return &allOutputTypes_[index];
+  }
+  const TensorShapeProto* getInputData(size_t index) const override {
+    if (index >= allInputData_.size()) {
+      ONNX_THROW("Input " + ONNX_NAMESPACE::to_string(index) + " is out of bounds.");
+    }
+    // Gets it from initializer if it exists
+    const auto* input_data = allInputData_[index];
+    if (input_data != nullptr) {
```

Comment:
also check for shape of input... allow only 1D or scalars

Reply:
Is that strictly necessary (for correctness)? We treat the returned value as representing the underlying sequence of data-values in the tensor, without saying anything about its shape. (The shape info, if any, is captured in the type.) I understand that in common use-cases this may not arise.

Reply:
Restricting this to 1D or scalars is okay (it should work for the common-case), but I still think it is not necessary. Allowing other cases enables a more general implementation without much of a cost. (But, given the deadline, I am okay with this restriction for now.)

Reply:
I suggest going with the restriction ... once we come across the models which require a more generic implementation we can update this...

Reply:
For now I still added the restriction since it's the most common case. Still, I added some comments here as TODO for future reference. Thanks for pointing out.